### ISSUE TITLE
Possible deadlock when client disconnect during http.response.body

### ISSUE THREAD

**Issue by eax64** [Kludex/uvicorn#336]
Hi,
It happens with both h11/httptools protocols and with both asyncio/uvloop loops.
I tried with uvicorn 0.6.1 on python 3.7.2, but I can also reproduce with python 3.6.8.

This issue doesn't occur every time. It seems to happen much less with log-level=debug.

app:
```python
import asyncio
import threading

_active_requests = 0
_lock = threading.Lock()


async def app(scope, receive, send):
    global _active_requests

    assert scope["type"] == "http"

    with _lock:
        _active_requests += 1

    print(f"Request started. Currently {_active_requests} active requests.")

    chunk_size = 1024*512
    chunks = [(str(i) * chunk_size).encode("utf-8") for i in range(10)]

    total_size = sum(len(c) for c in chunks)

    await send({
        "type": "http.response.start",
        "status": 200,
        "headers": [
            [b"content-type", b"text/plain"],
            [b"Content-Length", f"{total_size}".encode("utf-8")],
        ]
    })

    await asyncio.sleep(1)

    for i, chunk in enumerate(chunks):
        await send({
            "type": "http.response.body",
            "body": chunk,
            "more_body": i < len(chunks) - 1,
        })

    with _lock:
        _active_requests -= 1

    print(f"Request ended. Currently {_active_requests} active requests.")
```

Client:
```
$> curl -v http://127.0.0.1:8000 -s > /dev/null
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET / HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.64.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< date: Wed, 27 Mar 2019 23:43:29 GMT
< server: uvicorn
< content-type: text/plain
< content-length: 5242880
< 
{ [65536 bytes data]
* Connection #0 to host 127.0.0.1 left intact
$> curl http://127.0.0.1:8000 -s > /dev/null
$> curl http://127.0.0.1:8000 -s > /dev/null
```

Uvicorn output, all is good:
```
(venv-3.7.2) $ uvicorn async_app:app --http h11 --loop asyncio
INFO: Started server process [21186]
INFO: Waiting for application startup.
INFO: ASGI 'lifespan' protocol appears unsupported.
INFO: Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
Request started. Currently 1 active requests.
INFO: ('127.0.0.1', 37864) - "GET / HTTP/1.1" 200
Request ended. Currently 0 active requests.
Request started. Currently 1 active requests.
INFO: ('127.0.0.1', 37866) - "GET / HTTP/1.1" 200
Request ended. Currently 0 active requests.
Request started. Currently 1 active requests.
INFO: ('127.0.0.1', 37870) - "GET / HTTP/1.1" 200
Request ended. Currently 0 active requests.
```

Same thing but this time closing the client before the end:
```
$> curl -s http://127.0.0.1:8000 | head -c 1
$> curl -s http://127.0.0.1:8000 | head -c 1
$> curl -s http://127.0.0.1:8000 | head -c 1
```
```
(venv-3.7.2) $> uvicorn async_app:app --http h11 --loop asyncio
INFO: Started server process [21511]
INFO: Waiting for application startup.
INFO: ASGI 'lifespan' protocol appears unsupported.
INFO: Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
Request started. Currently 1 active requests.
INFO: ('127.0.0.1', 37938) - "GET / HTTP/1.1" 200
Request started. Currently 2 active requests.
INFO: ('127.0.0.1', 37940) - "GET / HTTP/1.1" 200
WARNING: socket.send() raised exception.
WARNING: socket.send() raised exception.
WARNING: socket.send() raised exception.
WARNING: socket.send() raised exception.
Request ended. Currently 1 active requests.
Request started. Currently 2 active requests.
INFO: ('127.0.0.1', 37942) - "GET / HTTP/1.1" 200
^CINFO: Shutting down
INFO: Waiting for background tasks to complete. (CTRL+C to force quit)
^CINFO: Finished server process [21511]
```

With more debug: 
```
$ uvicorn async_app:app --http h11 --loop asyncio --log-level debug
DEBUG: None - ASGI [1] Initialized {'type': 'lifespan'}
INFO: Started server process [21837]
INFO: Waiting for application startup.
DEBUG: None - ASGI [2] Initialized {'type': 'lifespan'}
DEBUG: None - ASGI [2] Started task
DEBUG: None - ASGI [2] Raised exception
INFO: ASGI 'lifespan' protocol appears unsupported.
INFO: Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
DEBUG: ('127.0.0.1', 37958) - Connected
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Initialized {'type': 'http', 'http_version': '1.1', 'server': ('127.0.0.1', 8000), 'client': ('127.0.0.1', 37958), 'scheme': 'http', 'method': 'GET', 'root_path': '', 'path': '/', 'query_string': b'', 'headers': '<...>'}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Started task
Request started. Currently 1 active requests.
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.start', 'status': 200, 'headers': '<...>'}
INFO: ('127.0.0.1', 37958) - "GET / HTTP/1.1" 200
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - ASGI [3] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37958) - Disconnected
DEBUG: ('127.0.0.1', 37964) - Connected
DEBUG: ('127.0.0.1', 37964) - ASGI [4] Initialized {'type': 'http', 'http_version': '1.1', 'server': ('127.0.0.1', 8000), 'client': ('127.0.0.1', 37964), 'scheme': 'http', 'method': 'GET', 'root_path': '', 'path': '/', 'query_string': b'', 'headers': '<...>'}
DEBUG: ('127.0.0.1', 37964) - ASGI [4] Started task
Request started. Currently 2 active requests.
DEBUG: ('127.0.0.1', 37964) - ASGI [4] Received {'type': 'http.response.start', 'status': 200, 'headers': '<...>'}
INFO: ('127.0.0.1', 37964) - "GET / HTTP/1.1" 200
DEBUG: ('127.0.0.1', 37964) - ASGI [4] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37964) - ASGI [4] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37964) - ASGI [4] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}
DEBUG: ('127.0.0.1', 37964) - Disconnected
^CINFO: Shutting down
INFO: Waiting for background tasks to complete. (CTRL+C to force quit)
^CINFO: Finished server process [21837]
```

strace:
```
strace -tt -e trace=network,read,write,epoll_wait,open,close -o strace-output -s 300 -f uvicorn ...
21912 23:54:22.101575 epoll_wait(3, [], 3, 0) = 0
21912 23:54:22.102320 write(2, "DEBUG: ('127.0.0.1', 37992) - ASGI [15] Received {'type': 'http.response.body', 'body
': '<524288 bytes>', 'more_body': True}\n", 125) = 125
21912 23:54:22.102798 sendto(7, "000000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"..., 524288, 0, N
ULL, 0) = 524288
21912 23:54:22.104034 write(2, "DEBUG: ('127.0.0.1', 37992) - ASGI [15] Received {'type': 'http.response.body', 'body
': '<524288 bytes>', 'more_body': True}\n", 125) = 125
21912 23:54:22.104506 sendto(7, "111111111111111111111111111111111111111111111111111111111111111111111111111111111111
111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111"..., 524288, 0, N
ULL, 0) = 491520
21912 23:54:22.105874 write(2, "DEBUG: ('127.0.0.1', 37992) - ASGI [15] Received {'type': 'http.response.body', 'body
': '<524288 bytes>', 'more_body': True}\n", 125) = 125
21912 23:54:22.108103 write(2, "DEBUG: ('127.0.0.1', 37992) - ASGI [15] Received {'type': 'http.response.body', 'body
': '<524288 bytes>', 'more_body': True}\n", 125) = 125
21912 23:54:22.108482 epoll_wait(3, [{EPOLLIN|EPOLLOUT|EPOLLERR|EPOLLHUP, {u32=7, u64=94802813124615}}], 3, 4) = 1
21912 23:54:22.108668 recvfrom(7, 0x56395ce668f0, 262144, 0, NULL, NULL) = -1 ECONNRESET (Connection reset by peer)
21912 23:54:22.109192 epoll_wait(3, [], 2, 0) = 0
21912 23:54:22.109539 write(2, "DEBUG: ('127.0.0.1', 37992) - Disconnected\n", 43) = 43
21912 23:54:22.109819 close(7)          = 0
21912 23:54:22.109952 epoll_wait(3, [], 2, 2) = 0
21912 23:54:22.112151 epoll_wait(3, [], 2, 0) = 0
21912 23:54:22.112325 epoll_wait(3, [], 2, 100) = 0
...
```

From what I've seen it seems it happens when:
- Something call pause_writing
- The ASGI send function is called and it's locked on `await self.flow.drain()`
- The connection_lost function is called.
- Nothing releases the lock and the ASGI send function never returns.

I've noticed this issue when investigating a memory leak with django/asgi/gunicorn. Clients are requesting some files (10MB - 200MB each) and some clients close the connection. The whole django response stays in memory forever since the asgi send function never return. This ends up taking all memory until uvicorn is restarted.

**Comment by lovelydinosaur** [comment]
Okay - I wonder if we ought to be calling `resume_writing` in `connection_lost()`, in case we're currently paused. Want to give that a try and see if it resolves the issue?

**Comment by eax64** [comment]
We're using uvloop plus httptools (through gunicorn) and calling `resume_writing` at the end of `connection_lost` seem to fix our memory leak.
```python
def connection_lost(self, exc):
        ...
        self.flow.resume_writing()
        self.message_event.set()
```

We do get this exception now though:
```
RuntimeError: unable to perform operation on <TCPTransport closed=True reading=False 0x2c6e9c8>; the handler is closed
  File "uvicorn/protocols/http/httptools_impl.py", line 379, in run_asgi
    result = await asgi(self.receive, self.send)
  File "channels/http.py", line 213, in __call__
    await self.handle(body)
  File "asgiref/sync.py", line 108, in __call__
    return await asyncio.wait_for(future, timeout=None)
  File "asyncio/tasks.py", line 339, in wait_for
    return (yield from fut)
  File "concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "asgiref/sync.py", line 123, in thread_handler
    return self.func(*args, **kwargs)
  File "channels/http.py", line 252, in handle
    self.send(response_message)
  File "asgiref/sync.py", line 64, in __call__
    return call_result.result()
  File "concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "asgiref/sync.py", line 78, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "uvicorn/protocols/http/httptools_impl.py", line 503, in send
    self.transport.write(body)
  File "uvloop/handles/stream.pyx", line 671, in uvloop.loop.UVStream.write
  File "uvloop/handles/handle.pyx", line 159, in uvloop.loop.UVHandle._ensure_alive
```


The same thing without django/channels but just the example app from the first comment gives this exception:
```
ERROR: Exception in ASGI application
Traceback (most recent call last):
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/protocols/http/httptools_impl.py", line 377, in run_asgi
    result = await asgi(self.receive, self.send)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/message_logger.py", line 63, in __call__
    raise exc from None
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/message_logger.py", line 59, in __call__
    await self.inner(self.receive, self.send)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/asgi3.py", line 12, in asgi
    await self.app(scope, receive, send)
  File "./async_app.py", line 38, in app
    "more_body": i < len(chunks) - 1,
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/message_logger.py", line 79, in send
    await self._send(message)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/protocols/http/httptools_impl.py", line 501, in send
    self.transport.write(body)
  File "uvloop/handles/stream.pyx", line 671, in uvloop.loop.UVStream.write
  File "uvloop/handles/handle.pyx", line 159, in uvloop.loop.UVHandle._ensure_alive
RuntimeError: unable to perform operation on <TCPTransport closed=True reading=False 0x558bfa334088>; the handler is closed
```

There doesn't seem to be any exception raised with httptools + asyncio, but the send function is not hanging anymore.

The exception with h11 for both asyncio/uvloop loops:
```
ERROR: Exception in ASGI application
Traceback (most recent call last):
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/protocols/http/h11_impl.py", line 380, in run_asgi
    result = await asgi(self.receive, self.send)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/message_logger.py", line 63, in __call__
    raise exc from None
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/message_logger.py", line 59, in __call__
    await self.inner(self.receive, self.send)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/asgi3.py", line 12, in asgi
    await self.app(scope, receive, send)
  File "./async_app.py", line 38, in app
    "more_body": i < len(chunks) - 1,
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/middleware/message_logger.py", line 79, in send
    await self._send(message)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/uvicorn/protocols/http/h11_impl.py", line 473, in send
    output = self.conn.send(event)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/h11/_connection.py", line 464, in send
    data_list = self.send_with_data_passthrough(event)
  File "/home/eax/.pyenv/versions/3.7.2/envs/venv-3.7.2/lib/python3.7/site-packages/h11/_connection.py", line 480, in send_with_data_passthrough
    "Can't send data when our state is ERROR")
h11._util.LocalProtocolError: Can't send data when our state is ERROR
```

Those exceptions might not be very helpful to understand what's happening. Maybe there is a way to properly cancel the response instead of letting the asgi app write more data even though the socket is closed?

**Comment by lovelydinosaur** [comment]
Sending the response to completion (but into the void, and without raising these exceptions) is the right way to tackle this.

We just need a little bit of rejigging around the “if disconnect occurs while writing is paused” case.

Those stacktraces are super helpful - thanks. Reckon I have enough to resolve this but will need to wait TIL tomorrow.

**Comment by lovelydinosaur** [comment]
Okay, so I think that #340 should resolve this - if you're able to test against that and confirm or otherwise that'd be ace.

**Comment by eax64** [comment]
It seems to work fine, no exception raised and no hanging connections.

FYI the strace for uvloop+httptools with additional debug print. You can see that the lock is cleared and that the asgi app still send the `http.response.body` messages after that.
```
8966  11:05:35.267335 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body'
: '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.267631 write(13, "000000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"..., 524288) = 52
4288
8966  11:05:35.268705 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body'
: '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.269070 write(13, "111111111111111111111111111111111111111111111111111111111111111111111111111111111111
111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111"..., 524288) = 65
483
8966  11:05:35.269457 write(13, "111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111"..., 458805) = -1 ECONNRESET (Connection reset by peer)
8966  11:05:35.269697 write(1, "DBG FlowControl.pause_writing()\n", 32) = 32
8966  11:05:35.270242 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.270501 write(1, "DBG FlowControl.drain(): Top\n", 29) = 29
8966  11:05:35.270770 epoll_wait(3, [{EPOLLIN|EPOLLHUP, {u32=13, u64=13}}], 1024, 0) = 1
8966  11:05:35.270996 read(13, "", 256000) = 0
8966  11:05:35.271248 write(1, "DBG HttpToolsProtocol.connection_lost(): Top\n", 45) = 45
8966  11:05:35.271560 write(2, "DEBUG: ('127.0.0.1', 46998) - Disconnected\n", 43) = 43
8966  11:05:35.271720 write(1, "DBG FlowControl.resume_writing()\n", 33) = 33
8966  11:05:35.271936 write(1, "DBG HttpToolsProtocol.connection_lost(): Bottom\n", 48) = 48
8966  11:05:35.272105 close(13)         = 0
8966  11:05:35.272217 epoll_wait(3, [], 1024, 0) = 0
8966  11:05:35.272362 write(1, "DBG FlowControl.drain(): Bottom\n", 32) = 32
8966  11:05:35.272883 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.273223 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.273631 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.273960 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.274198 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.274581 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': True}\n", 124) = 124
8966  11:05:35.274922 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Received {'type': 'http.response.body', 'body': '<524288 bytes>', 'more_body': False}\n", 125) = 125
8966  11:05:35.275105 write(1, "Request ended. Currently 0 active requests.\n", 44) = 44
8966  11:05:35.276206 write(2, "DEBUG: ('127.0.0.1', 46998) - ASGI [5] Completed\n", 49) = 49
```

Thanks!

**Comment by lovelydinosaur** [comment]
Great - that's as it should be.
Letting the server complete the response normally is the most graceful pattern here.

**Comment by lovelydinosaur** [comment]
Okay, released as 0.7.0.b2.  You'll want to install with `pip install --pre` since it's labelled as a beta pre-release.

**Comment by lovelydinosaur** [comment]
Thanks for helping work through this!
